### PR TITLE
Datasource: datasource config page header redesign

### PR DESCRIPTION
**Pull request by ktw4071**
**What is this feature?**

Datasource config page header redesign

**Why do we need this feature?**

Datasource overhaul, redesigning to improve user experience

**Who is this feature for?**

Datasource plugin teams

**Which issue(s) does this PR fix?**:

Fixes #
https://github.com/grafana/enterprise-datasources/issues/156

## checklist
- [x] removal of the breadcrumbs from the name and making it editable (see figma file for edit states: https://www.figma.com/file/mijj01wckVu4xrZ18JOVTf/Configuration-Redesign?node-id=923:58843&t=KJHgRsCknNv4JllN-1)
- [x] badge for supported features (currently `Alerting`)
- [x] ~data source delete button~
- [x] ~explore data/view dashboards buttons visible after a successful save < note: change this to visible at all times to keep consistency and consistency w/ the connections console~
- [x] using button group for view dahsboard, explore, delete
- [x] default data source toggle
- [x] removing name field from the form to not duplicate the functionality
- [x] removing explore button in the footer
- [x] styling on redesign

## Before change
image

## After change
image

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by catherineymgui** [comment]
- I noticed these elements of the header are now persisting across all of grafana headers?

Screenshot 2023-04-25 at 17 11 41
Screenshot 2023-04-25 at 17 11 36

- Interaction pattern should be: user can click enter or click the checkmark or leave out of the input field, and it saves. Once it saves, you exit the input field all together and it just returns to the state with the edit pencil. Right now it only saves on the “save and test” button and the input field persists.

- There should be a checkmark icon button beside the input field
Screenshot 2023-04-25 at 17 12 10

- In the designs, the name is truncated, which was why I was wondering if the input field should be a fixed width. If we keep the responsive header and the name just gets longer, a better input field interaction would be that it would be the input field getting bigger with the name (with a min width to align to the alerting badge)

- Input field should have min width to align with the alerting supported text
Screenshot 2023-04-25 at 17 12 54

- Make sure the right hand buttons have 8px spacing between them (is it 16px right now?)
- Make sure there’s 16px padding between the type: and the default field
- Make sure the alerting badge is aligned to the right of the container
- Make sure the name of the type is using the primary text colour

File: public/app/features/datasources/components/EditDataSourceTitle.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDataSourceTitle.tsx:33]: no need for this, you can set autoFocus={true} on the input

File: public/app/features/datasources/components/EditDataSourceTitle.tsx
Comment (torkelo) [reply public/app/features/datasources/components/EditDataSourceTitle.tsx:54]: Also this input does auto-expand as I type

File: public/app/features/datasources/components/EditDataSourceTitle.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDataSourceTitle.tsx:82]: this should use the theme h1 fontSize

File: public/app/features/datasources/components/EditDatasSourceSubtitle.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDatasSourceSubtitle.tsx:51]: This makes this code very fragile to how we format the subtitle in useDataSourceSettingsNav. please pass the plugin name instead via props .

File: public/app/features/datasources/components/EditDatasSourceSubtitle.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDatasSourceSubtitle.tsx:36]: Getting a warning here due to uncontrolled input, not sure why really. isDefault has a value here 
image

File: public/app/features/datasources/components/EditDataSourceActions.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDataSourceActions.tsx:59]: This does not go to "View dashboards" page but a new dashboard with a a new panel, with this data source instance pre-selected

**Review by torkelo — COMMENTED** [review]
Looking better, not super happy with the design (left a proposal in the project channel). Still some issues with the code but getting closer :)

**Comment by sasklacz** [comment]
> @ktw4071 just one small comment to remember
> 
> > will have to work on alignment of components in the follow-up PR
> 
> Main branch must always be releasable - we have environments on Grafana Cloud that are continiously deployed to from main every hour, so work must be sequenced so main is never 'broken' or 'incomplete'. Work-in-progress work should be gated behind feature flags if it can't go into main "complete".

I think in this case this is supposed to be available immediately for all plugins.

File: public/app/features/connections/pages/EditDataSourcePage.tsx
Comment (leventebalogh) [reply public/app/features/connections/pages/EditDataSourcePage.tsx:29]: Yeah, just like Torkel says I think it would make sense to keep these "implementation details" in the sub-components instead (in `<EditDataSourceSubtitle>` and in `<EditDataSourceTitle>`), because in that case the page component would remain "leaner" and the sub-components would be easier to reuse / move around.

This would also be helpful because these changes will also need to be "mirrored" to [connections/pages/EditDataSourcePage](https://github.com/grafana/grafana/blob/main/public/app/features/connections/pages/EditDataSourcePage.tsx), and like that we won't duplicate this logic.

File: public/app/features/datasources/components/EditDatasSourceSubtitle.tsx
Comment (leventebalogh) [inline public/app/features/datasources/components/EditDatasSourceSubtitle.tsx:1]: We introduced `EditDataSourceSubtitle.tsx` and `EditDataSourceTitle.tsx` while removed `BasicSettings.tsx` - this can be fine, however while the basic settings had tests the new sub-components don't.

Let's add tests for these (either for the sub-components or by testing them through the `EditDataSourcePage`).

**Comment by torkelo** [comment]
don't merge this without approval from me please

**Comment by catherineymgui** [comment]
Some styling things but it might be easier to do this at once with a new header design? (let's chat tmrw):

- [ ] Update the edit icon to be the secondary text colour
- [ ] Can the edit icon be aligned to the bottom and beside the header text (ie like this) 
Screenshot 2023-04-27 at 23 07 29

- [ ] Let's have the alert badge beside the default icon (16px) instead of right aligned 
- [ ] Have the Type: Datasource font size also 12px
- [ ] Can there be 4px padding b/w the 'Default' and info icon, and 8px b/w info icon and the toggle?
- [ ] Update default tooltip to be: "The default data source is preselected in new panels"

**Review by torkelo — CHANGES_REQUESTED** [review]
* The header changes completely when moving to another tab. The header should stay the same no matter what tab is active, having the buttons and subtitle change when you switch tab is very confusing/distracting. 

image

I think this needs some more design thought. Using my proposed design is more consistent with the plugin details page (looks a lot better, and feels more harmonious). And more importantly, would make the design work much better for all tabs. 

Suggest
* Move Delete button to the bottom of settings page
* Use page info items for data source type & alerting supported state
* Move build a dashboard and explore to be normal links in page description. Free up space for smaller screens and long data source names. 
* The default toggle think works best as a page action item (info item) instead of having form options in the description. And think works best this way given this will be here for all tabs.  
* Add title (tooltip) text to edit title pen button

* 

Screenshot from 2023-04-27 08-42-37

Some bugs
* All actions are disabled for a provisioned data source
* Edit data source name pen button is available on read-only data sources (provisioned data sources) 
* Editing a data source name via pen button should save it since this is outside the Settings tab / form. (and needs to be available / work on all tabs)

**Comment by leventebalogh** [comment]
Two more things that I found while taking the branch for spin:

- the name-change is not persisted (not even for non-provisioned datasources)
- setting a datasource as default (with the switch) is not persisted
- the little info-icon doesn't show a tooltip in some cases 
- clicking the little info icon next to the default switch is toggling the switch

https://user-images.githubusercontent.com/9974811/235120448-5f6fc03e-6c67-464a-a4df-61fd8df185d0.mov

File: public/app/features/datasources/components/useDataSourceInfo.tsx
Comment (torkelo) [inline public/app/features/datasources/components/useDataSourceInfo.tsx:35]: Please use `<Badge color="green" ... ` here

File: public/app/features/connections/pages/EditDataSourcePage.tsx
Comment (torkelo) [inline public/app/features/connections/pages/EditDataSourcePage.tsx:38]: why this? Is it to remove the description line that has "Type: Prometheus", can't we remove that where the nav model is built instead?

File: public/app/features/connections/pages/EditDataSourcePage.tsx
Comment (torkelo) [inline public/app/features/connections/pages/EditDataSourcePage.tsx:24]: instead of this abstraction, I would just make a wrapper component that owns setting up the Page component (and passes in navId as props). DataSourceTabPage or something (so we can easily have the same header for all the tabs)

**Review by torkelo — COMMENTED** [review]
This is starting to look really good from my POV. 

My only complaint is that the header changes radically when switching tabs, reverting to the old plain header for all but the settings tab. I know it's sort of like that in main, think this was a miss when we added the actions buttons to the header.  

We do not need to fix that in this PR but I think it's an important follow-up to address.

**Comment by catherineymgui** [comment]
Here are some small things I found. I'll do another round once everything is addressed!

Default toggle:
Just a fyi - This won’t be needed for panel edit anymore — they will just keep the preselected data source instead of a “default”. Confirming whether this is still needed for explore.

- [ ] If not: We can remove this section
- [ ] If yes: Is it possible to add an info icon w/ tooltip? 
- [ ] If yes: Is it possible to tighten the padding b/w the toggle and the ‘Default’ label?
- [ ] If yes: Can we add a disabled state / and tooltip when it is disabled (when the data source was added by config)

Alerting badge
- [ ] If you hover — the cursor changes implying a hover state / that you can click it. This we remove this state?

Margin
- [ ] Can we add 24px margin b/w the types and the buttons (instead of 16)

File: public/app/features/connections/pages/EditDataSourcePage.tsx
Comment (torkelo) [inline public/app/features/connections/pages/EditDataSourcePage.tsx:12]:
```suggestion
  return <DataSourceTabPage uid={uid} pageId={pageId} navId="connections-datasources" />;
```

thin this navId has changed

**Review by torkelo — COMMENTED** [review]
The default info icon does nothing for me 

Screenshot 2023-05-05 at 16 34 30

**Comment by leventebalogh** [comment]
I think it's getting there, I have a few remarks though:

**The info tooltip for the deafult-switch doesn't seem to work for me:**
https://user-images.githubusercontent.com/9974811/236785668-a2fed34a-753f-477f-9462-9953a623a56d.mov

**Not working default switch**
As @torkelo  pointed it out, too, it doesn't work in cases when the datasource form below has any errors. It is not _strictly_ related to this change, however as the default-switch and the name-input has been lifted out the form itself, now users will understand less why they are not working in case of errors. (While they were part of the form it was more intuitive to highlight the errors.) Same with the name change - it only works if the datasource modal doesn't have any errors. **The main problem is that as we have moved the form errors to the bottom of the form and removed the notifications, the user now has no feedback about the errors.**


https://user-images.githubusercontent.com/9974811/236786890-d29e184a-b3af-4497-9d9b-46e02671cae2.mov



I would say that this is a major UX issue, that we should address somehow.

**Comment by shelldandy** [comment]
image

Added ellipsis on the title at 500px wide max

**Comment by leventebalogh** [comment]
> However, how about we make API available for updating field change that does not rely on validation on other fields?

Yeah, I think probably that would be the best way forward, as I don't see an easy way for showing the errors for the header actions (without bringing back the error notifications). I will have a think today about how we could do it.

File: public/app/features/datasources/components/EditDataSourceActions.tsx
Comment (leventebalogh) [inline public/app/features/datasources/components/EditDataSourceActions.tsx:55]: I am not sure about this change - did we confirm that we are losing tracking of events because of navigating away? I think Rudderstack implemented using `navigator.sendBeacon()`, so those requests won't get cancelled once the page is navigating away.)

My concern with this change is that it stops behaving like a regular link, people cannot Ctrl+Click / Cmd+Click it anymore, or Right-click -> Open in new tab.

File: public/app/features/datasources/components/EditDataSourceTitle.tsx
Comment (leventebalogh) [inline public/app/features/datasources/components/EditDataSourceTitle.tsx:11]: As it doesn't hold any editing anymore, let's rename it to `DataSourceTitle`?

File: public/app/features/datasources/state/navModel.ts
Comment (leventebalogh) [inline public/app/features/datasources/state/navModel.ts:19]: Just for confirmation - are we sure we would like to remove the "Type: Prometheus" as a subtitle from the datasources header? (This is what tells the user what type of datasource they are looking at - besides the logo).

File: public/app/features/datasources/types.ts
Comment (leventebalogh) [inline public/app/features/datasources/types.ts:23]: As it's only the `useDataSourceInfo()` hook using this type we could just move it in `useDataSourceInfo.tsx`?

**Comment by leventebalogh** [comment]
Took it for a spin, and it all looks working. Left a few comments, mostly minor, only the buttons used as links makes me concerned. Let's wait for @torkelo to have a look as well.

File: public/app/features/datasources/components/EditDataSourceActions.tsx
Comment (torkelo) [inline public/app/features/datasources/components/EditDataSourceActions.tsx:37]: why not a normal link with onClick that just handles tracking?

File: public/app/features/datasources/components/EditDataSourceActions.tsx
Comment (torkelo) [reply public/app/features/datasources/components/EditDataSourceActions.tsx:55]: LinkButton should support onClick so not sure why we need to change it to Button

File: public/app/features/datasources/state/navModel.ts
Comment (catherineymgui) [reply public/app/features/datasources/state/navModel.ts:19]: While it's not a subtitle, it's still showing up as part of the page info. Is there a concern with having it as a page info vs subtitle?

This is mostly to match the new page header component. I agree that if "type" was the only piece of info it could live elsewhere.

Comment (leventebalogh) [reply public/app/features/datasources/state/navModel.ts:19]: Sorry, I somehow missed that it is part of the info section now. No, I think it's fine if we have it in there 👍

**Comment by leventebalogh** [comment]
How are we planning we deploy the feature-toggle?

Looks good to me, although let's wait until @torkelo has a look at it as well.

**Review by torkelo — APPROVED** [review]
Think this looks good, biggest thing missing which we need to create follow up issues for is to make sure this page header is shown on all tabs. Having the page header change when switching tabs feels pretty bad

**Review by leventebalogh — APPROVED** [review]

**Review by sasklacz — APPROVED** [review]
Agree with what Torkel said about header switching between the tabs. Maybe we can talk this through during our sync on Wednesday

**Pull request merged**
